### PR TITLE
Add pyupgrade to pre-commit

### PR DESCRIPTION
**Pull request by dbast**
### Description

Adding and applying pyupgrade (same upstream as pre-commit):
* to remove superfluous future imports
* convert various old string formattings to fstrings
* replace mock imports with integrated unittest.mock
* Update aliases of OSError to OSError
* ... for more see [pyupgrade](https://github.com/asottile/pyupgrade)

conda-build and conda-pack have this already applied.

This requires pre-commit to apply twice (due to string formatting upgrades) and a small commit on top to remove mock import cascades.

Supersedes #11903

### Checklist - did you ...

- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

**Comment by dbast** [comment]
pre-commit.ci autofix

**Comment by dbast** [comment]
pre-commit.ci autofix

**Comment by dbast** [comment]
@kenodegard Sorry did not see your PR https://github.com/conda/conda/pull/11903 

This PR contains exactly 4 commits: The initial pre-commit activation, 2 pre-commit auto-fix commits, 1 manual commit at the end to fix a few remaining flake8 findings and removing the import cascades.

**Comment by dbast** [comment]
conda/exports.py is excluded from further pyupgrade runs as that would remove `from builtins import input`, which is imported by conda-build(!).

**Comment by dbast** [comment]
rebased due to changes on main.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -32,6 +32,12 @@ repos:
       - id: fix-encoding-pragma
         args: [--remove]
         exclude: ^conda/auxlib/
+  - repo: https://github.com/asottile/pyupgrade
+    rev: v2.38.2
+    hooks:
+      - id: pyupgrade
+        args: ["--py37-plus", "--keep-runtime-typing"]
```
Comment (kenodegard) [inline .pre-commit-config.yaml:39]: I don't think we need to preserve the typing?

```suggestion
        args: ["--py37-plus"]
```

**Review by kenodegard — COMMENTED** [review]
`pyupgrade` is magical

File: .pre-commit-config.yaml (hunk 1 above)
Comment (dholth) [reply .pre-commit-config.yaml]: By removing `--keep-runtime-typing` we'll get types like `str | None` or `list[str]` that aren't parsed at runtime. It is annoying when pyupgrade edits your carefully typed types but it is more modern.

**Comment by dholth** [comment]
I felt a great disturbance in the pull requests, as if a thousand good merges turned bad

**Comment by dbast** [comment]
@dholth would you feel better if we disable the f-string conversion (have to test what the `--keep-percent-format` pyupgrade option does)? that would potentially touch much less existing code.

Splitting the PR is also possible: mock removal, futures removal, object removal, exception alias removal... and maybe f-string conversion later.

What do we feel most comfortable with?

**Comment by dbast** [comment]
pre-commit.ci autofix

**Review by kenodegard — APPROVED** [review]
@dbast thanks!

**Review by jezdez — APPROVED** [review]
Approving the configuration, not the automated code changes, since I'm trusting pyupgrade to do what it says on the tin.

File: conda/common/_logic.py [hunk 2]
```diff
@@ -16,7 +13,7 @@
 FALSE = -TRUE
```
Comment (dholth) [inline conda/common/_logic.py:13]: 🙈

File: conda/common/_logic.py [hunk 3]
```diff
@@ -255,15 +252,15 @@ def process_solution(self, sat_solution):
...
+class Clauses:
     def __init__(self, m=0, sat_solver_str=_sat_solver_cls_to_str[_PycoSatSolver]):
         self.unsat = False
         self.m = m
 
         try:
             sat_solver_cls = _sat_solver_str_to_cls[sat_solver_str]
         except KeyError:
-            raise NotImplementedError("Unknown SAT solver: {}".format(sat_solver_str))
+            raise NotImplementedError(f"Unknown SAT solver: {sat_solver_str}")
```
Comment (dholth) [inline conda/common/_logic.py:263]: perhaps the SAT solver is unsatisfactory

File: conda/core/prefix_data.py [hunk 4]
```diff
@@ -352,7 +350,7 @@ def get_conda_anchor_files_and_records(site_packages_short_path, python_records)
     conda_python_packages = odict()
 
     matcher = re.compile(
-        r"^%s/[^/]+(?:%s)$" % (
+        r"^{}/[^/]+(?:{})$".format(
```
Comment (dholth) [inline conda/core/prefix_data.py:353]: some `.format` remains; doesn't bother me but interesting

File: conda/gateways/disk/create.py [hunk 5]
```diff
@@ -248,20 +246,22 @@ def make_menu(prefix, file_path, remove=False):
 
 def create_hard_link_or_copy(src, dst):
     if islink(src):
-        message = dals("""
+        message = dals(
+            """
         Cannot hard link a soft link
```
Comment (dholth) [inline conda/gateways/disk/create.py:251]: you WOULDN't
    hardLINK a 
SOFT link

File: conda/resolve.py [hunk 6]
```diff
@@ -134,15 +133,16 @@ def __init__(self, index, processed=False, channels=()):
...
-                )
+        return (
+            super().__hash__()
+            ^ hash(frozenset(self.channels))
+            ^ hash(frozendict(self._channel_priorities_map))
+            ^ hash(self._channel_priority)
+            ^ hash(self._solver_ignore_timestamps)
+            ^ hash(frozendict((k, tuple(v)) for k, v in self.groups.items()))
+            ^ hash(frozendict((k, tuple(v)) for k, v in self.trackers.items()))
+            ^ hash(frozendict((k, tuple(v)) for k, v in self.ms_depends_.items()))
```
Comment (dholth) [inline conda/resolve.py:144]: looks expensive

File: conda/testing/decorators.py [hunk 7]
```diff
@@ -1,24 +0,0 @@
...
-    except ImportError:
-        skip_mock = True
-
-
-def skip_if_no_mock(func):
-    @wraps(func)
-    @unittest.skipIf(skip_mock, 'install mock library to test')
-    def inner(*args, **kwargs):
-        return func(*args, **kwargs)
-    return inner
```
Comment (dholth) [inline conda/testing/decorators.py:24]: solid changeset

File: tests/conda_env/test_cli.py [hunk 8]
```diff
@@ -175,11 +175,11 @@ def run_conda_command(command, prefix, *arguments):
...
         arguments = list(map(escape_for_winpath, arguments))
     if command is Commands.INFO:    # INFO
-        command_line = "{0} {1}".format(command, " ".join(arguments))
+        command_line = "{} {}".format(command, " ".join(arguments))
     elif command is Commands.LIST:  # LIST
-        command_line = "{0} -n {1} {2}".format(command, prefix, " ".join(arguments))
+        command_line = "{} -n {} {}".format(command, prefix, " ".join(arguments))
     else:  # CREATE
-        command_line = "{0} -y -q -n {1} {2}".format(command, prefix, " ".join(arguments))
+        command_line = "{} -y -q -n {} {}".format(command, prefix, " ".join(arguments))
```
Comment (dholth) [inline tests/conda_env/test_cli.py:182]: additional `.format`

**Review by dholth — APPROVED** [review]
LGTM

Needs NEWS explaining (tested) steps to merge into all the existing PR's

**Comment by kenodegard** [comment]
@dholth Good idea to add this to the skip blame after merging

I've added a news snippet

Unsure where to add additional rebasing instructions, I just tried rebasing several PRs, and the standard rebase (not merge pull) worked splendidly.

**Review by dholth — COMMENTED** [review]
expand pyupgrade explanation

**Comment by kenodegard** [comment]
On the development branch I tested both rebase and merge resulted in 5 conflicts needing manual fixes. Either way, these conflicts were easy to work through, so I don't see much of an issue merging this. Instructions are standard PR instructions to rebase/merge your branch as necessary.

I advocate for rebasing PRs to keep the commit history clean because there's usually only one user spearheading the PR. I also find rebasing less confusing if/when conflicts occur as it replays my changes. Just a personal preference 🤷🏼‍♂️

**Comment by dholth** [comment]
And we squash merge which is another way to keep the global history clean.

File: news/11909-pyupgrade [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+* <news item>
+
+### Docs
+
+* <news item>
+
+### Other
+
+* Dropped pragma fixes from pre-commit checks. (#11909)
+* Added pyupgrade to pre-commit checks. (#11909)
```
Comment (dholth) [inline news/11909-pyupgrade:20]: * Added pyupgrade to pre-commit checks. (#11909). This change affects many
  files. Existing pull requests may need to be updated, rebased, or merged to
  avoid conflicts.

Comment (kenodegard) [reply news/11909-pyupgrade:20]:
```suggestion
* Added pyupgrade to pre-commit checks. This change affects many
  files. Existing pull requests may need to be updated, rebased, or merged to
  address conflicts. (#11909)
```

**Review by dholth — APPROVED** [review]

**Pull request merged**